File: update-readme.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");

class HomebrewTapUpdater {
  constructor() {
    this.casksDir = path.join(__dirname, "Casks");
    this.formulaDir = path.join(__dirname, "Formula");
    this.readmePath = path.join(__dirname, "README.md");
    this.dryRun = process.argv.includes("--dry-run");
  }

  // Parse Cask file
  parseCaskFile(filePath) {
    try {
      const content = fs.readFileSync(filePath, "utf8");
      const fileName = path.basename(filePath, ".rb");

      // Get file last modified time
      const stats = fs.statSync(filePath);
      const lastModified = stats.mtime;

      // Extract version number
      const versionMatch = content.match(/version\s+["']([^"']+)["']/);
      const version = versionMatch ? versionMatch[1] : "Unknown";

      // Extract application name
      const nameMatch = content.match(/name\s+["']([^"']+)["']/);
      const name = nameMatch ? nameMatch[1] : this.formatAppName(fileName);

      // Extract description
      const descMatch = content.match(/desc\s+["']([^"']+)["']/);
      const desc = descMatch ? descMatch[1] : "No description available";

      return {
        fileName,
        name,
        desc,
        version,
        type: "GUI App",
        lastModified,
      };
    } catch (error) {
      console.error(`Error parsing cask file ${filePath}:`, error.message);
      return null;
    }
  }

  // Parse Formula file
  parseFormulaFile(filePath) {
    try {
      const content = fs.readFileSync(filePath, "utf8");
      const fileName = path.basename(filePath, ".rb");

      // Get file last modified time
      const stats = fs.statSync(filePath);
      const lastModified = stats.mtime;

      // Extract version number
      const versionMatch = content.match(/version\s+["']([^"']+)["']/);
      const version = versionMatch ? versionMatch[1] : "Unknown";

      // Extract class name as application name
      const classMatch = content.match(/class\s+(\w+)\s+<\s+Formula/);
      const name = classMatch ? classMatch[1].toLowerCase() : fileName;

      // Extract description
      const descMatch = content.match(/desc\s+["']([^"']+)["']/);
      const desc = descMatch ? descMatch[1] : "No description available";

      return {
        fileName,
        name,
        desc,
        version,
        type: "CLI Tool",
        lastModified,
      };
    } catch (error) {
      console.error(`Error parsing formula file ${filePath}:`, error.message);
      return null;
    }
  }

  // Format application name
  formatAppName(fileName) {
    return fileName
      .split("-")
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
      .join(" ");
  }

  // Get all application information
  getAllApps() {
    const apps = [];

    // Parse Casks
    if (fs.existsSync(this.casksDir)) {
      const caskFiles = fs
        .readdirSync(this.casksDir)
        .filter((file) => file.endsWith(".rb"));

      for (const file of caskFiles) {
        const filePath = path.join(this.casksDir, file);
        const appInfo = this.parseCaskFile(filePath);
        if (appInfo) {
          apps.push(appInfo);
        }
      }
    }

    // Parse Formula
    if (fs.existsSync(this.formulaDir)) {
      const formulaFiles = fs
        .readdirSync(this.formulaDir)
        .filter((file) => file.endsWith(".rb"));

      for (const file of formulaFiles) {
        const filePath = path.join(this.formulaDir, file);
        const appInfo = this.parseFormulaFile(filePath);
        if (appInfo) {
          apps.push(appInfo);
        }
      }
    }

    // Sort by last modified time (newest first)
    return apps.sort((a, b) => b.lastModified - a.lastModified);
  }

  // Generate table rows
  generateTableRows(apps) {
    return apps
      .map((app) => {
        const version = app.version.startsWith("v")
          ? app.version
          : `v${app.version}`;
        return `| ${app.name} | ${app.type} | ${app.desc} | ${version} |`;
      })
      .join("\n");
  }

  // Update README.md
  updateReadme() {
    try {
      if (!fs.existsSync(this.readmePath)) {
        console.error("README.md file not found!");
        return false;
      }

      const apps = this.getAllApps();
      console.log(`Found ${apps.length} applications:`);

      apps.forEach((app) => {
        console.log(`  - ${app.name} (${app.type}): v${app.version}`);
      });

      const readmeContent = fs.readFileSync(this.readmePath, "utf8");

      // Find table start and end positions
      const tableStartRegex =
        /\| Application Name\s+\| Type\s+\| Description\s+\| Latest Version\s+\|/;
      const tableHeaderSeparator = /\|\s*-+\s*\|\s*-+\s*\|\s*-+\s*\|\s*-+\s*\|/;

      const startMatch = readmeContent.match(tableStartRegex);
      if (!startMatch) {
        console.error("Could not find application table in README.md");
        return false;
      }

      const startIndex = startMatch.index;
      const headerEndIndex = readmeContent.indexOf("\n", startIndex);
      const separatorEndIndex = readmeContent.indexOf("\n", headerEndIndex + 1);

      // Find table end position (next ## title or end of file)
      const afterTable = readmeContent.substring(separatorEndIndex + 1);
      const nextSectionMatch = afterTable.match(/\n## /);
      const tableEndIndex = nextSectionMatch
        ? separatorEndIndex + 1 + nextSectionMatch.index
        : readmeContent.length;

      // Build new README content
      const beforeTable = readmeContent.substring(0, separatorEndIndex + 1);
      const afterTableContent = readmeContent.substring(tableEndIndex);
      const newTableRows = this.generateTableRows(apps);

      const newContent = beforeTable + newTableRows + "\n" + afterTableContent;

      if (this.dryRun) {
        console.log("\n=== DRY RUN MODE ===");
        console.log("New table content would be:");
        console.log(newTableRows);
        console.log("\n=== END DRY RUN ===");
        return true;
      }

      fs.writeFileSync(this.readmePath, newContent, "utf8");
      console.log("\n✅ README.md updated successfully!");

      return true;
    } catch (error) {
      console.error("Error updating README.md:", error.message);
      return false;
    }
  }

  // Check version consistency
  checkVersionConsistency() {
    console.log("\n🔍 Checking for version inconsistencies...");

    const apps = this.getAllApps();
    const readmeContent = fs.readFileSync(this.readmePath, "utf8");

    let hasInconsistencies = false;

    apps.forEach((app) => {
      const versionInReadme = this.extractVersionFromReadme(
        readmeContent,
        app.name
      );
      if (versionInReadme && versionInReadme !== `v${app.version}`) {
        console.log(
          `⚠️  ${app.name}: README shows ${versionInReadme}, actual is v${app.version}`
        );
        hasInconsistencies = true;
      }
    });

    if (!hasInconsistencies) {
      console.log("✅ All versions are consistent!");
    }

    return !hasInconsistencies;
  }

  // Extract version number from README
  extractVersionFromReadme(content, appName) {
    const regex = new RegExp(
      `\\|\\s*${appName}\\s*\\|[^\\|]*\\|[^\\|]*\\|\\s*([^\\|\\s]+)\\s*\\|`,
      "i"
    );
    const match = content.match(regex);
    return match ? match[1].trim() : null;
  }

  // Run the updater
  run() {
    console.log("🚀 Starting Homebrew Tap README updater...");

    if (this.dryRun) {
      console.log("📋 Running in DRY RUN mode - no files will be modified");
    }

    // Check version consistency
    this.checkVersionConsistency();

    // Update README
    const success = this.updateReadme();

    if (success) {
      console.log("\n🎉 Update completed successfully!");
    } else {
      console.log("\n❌ Update failed!");
      process.exit(1);
    }
  }
}

// Run the updater
if (require.main === module) {
  const updater = new HomebrewTapUpdater();
  updater.run();
}

module.exports = HomebrewTapUpdater;

File: scripts/update-readme.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "optparse"
require "open3"
require "pathname"
require_relative "lib/homebrew_tap/package_tools"

options = { dry_run: false }

OptionParser.new do |parser|
  parser.banner = "Usage: scripts/update-readme.rb [--dry-run]"
  parser.on("--dry-run", "Print the generated table without writing README.md") { options[:dry_run] = true }
end.parse!

class ReadmeUpdater
  TABLE_HEADER = /\| Application Name\s+\| Type\s+\| Description\s+\| Latest Version\s+\|/.freeze

  def initialize(root:, dry_run:)
    @root = Pathname.new(root)
    @dry_run = dry_run
    @readme_path = @root.join("README.md")
  end

  def run
    puts "Starting Homebrew Tap README updater"
    puts "Dry run: README.md will not be modified" if @dry_run

    check_version_consistency
    update_readme
    puts "Update completed successfully"
  end

  private

  def update_readme
    raise "README.md file not found" unless @readme_path.file?

    apps = all_apps
    puts "Found #{apps.length} applications:"
    apps.each { |app| puts "  - #{app.fetch(:name)} (#{app.fetch(:type)}): #{display_version(app.fetch(:version))}" }

    content = @readme_path.read
    start_match = content.match(TABLE_HEADER)
    raise "Could not find application table in README.md" unless start_match

    header_end = content.index("\n", start_match.begin(0))
    separator_end = content.index("\n", header_end + 1)
    raise "Could not find application table separator in README.md" unless separator_end

    after_table = content[(separator_end + 1)..]
    next_section = after_table&.match(/\n## /)
    table_end = next_section ? separator_end + 1 + next_section.begin(0) : content.length

    new_rows = table_rows(apps)
    new_content = content[0..separator_end] + new_rows + "\n" + content[table_end..]

    if @dry_run
      puts "\n=== DRY RUN MODE ==="
      puts "New table content would be:"
      puts new_rows
      puts "=== END DRY RUN ==="
    else
      @readme_path.write(new_content)
      puts "README.md updated successfully"
    end
  end

  def all_apps
    casks = @root.join("Casks").glob("*.rb").map { |path| parse_cask(path) }.compact
    formulae = @root.join("Formula").glob("*.rb").map { |path| parse_formula(path) }.compact

    (casks + formulae).sort_by { |app| [-app.fetch(:last_modified), app.fetch(:name).downcase] }
  end

  def parse_cask(path)
    content = path.read
    file_name = path.basename(".rb").to_s
    {
      file_name: file_name,
      name: match(content, /name\s+["']([^"']+)["']/) || format_app_name(file_name),
      desc: match(content, /desc\s+["']([^"']+)["']/) || "No description available",
      version: match(content, /version\s+["']([^"']+)["']/) || "Unknown",
      type: "GUI App",
      last_modified: git_modified_time(path)
    }
  rescue StandardError => e
    warn "Error parsing cask file #{path}: #{e.message}"
    nil
  end

  def parse_formula(path)
    content = path.read
    file_name = path.basename(".rb").to_s
    class_name = match(content, /class\s+(\w+)\s+<\s+Formula/)
    {
      file_name: file_name,
      name: class_name ? class_name.downcase : file_name,
      desc: match(content, /desc\s+["']([^"']+)["']/) || "No description available",
      version: match(content, /version\s+["']([^"']+)["']/) || "Unknown",
      type: "CLI Tool",
      last_modified: git_modified_time(path)
    }
  rescue StandardError => e
    warn "Error parsing formula file #{path}: #{e.message}"
    nil
  end

  def check_version_consistency
    puts "\nChecking for version inconsistencies..."

    content = @readme_path.read
    inconsistent = false
    all_apps.each do |app|
      readme_version = extract_version_from_readme(content, app.fetch(:name))
      actual_version = display_version(app.fetch(:version))
      next unless readme_version && readme_version != actual_version

      puts "  #{app.fetch(:name)}: README shows #{readme_version}, actual is #{actual_version}"
      inconsistent = true
    end

    puts "All versions are consistent" unless inconsistent
  end

  def table_rows(apps)
    apps.map do |app|
      "| #{app.fetch(:name)} | #{app.fetch(:type)} | #{app.fetch(:desc)} | #{display_version(app.fetch(:version))} |"
    end.join("\n")
  end

  def extract_version_from_readme(content, app_name)
    escaped = Regexp.escape(app_name)
    match(content, /\|\s*#{escaped}\s*\|[^|]*\|[^|]*\|\s*([^|\s]+)\s*\|/i)
  end

  def display_version(version)
    version.to_s.start_with?("v") ? version.to_s : "v#{version}"
  end

  def format_app_name(file_name)
    file_name.split("-").map(&:capitalize).join(" ")
  end

  def git_modified_time(path)
    relative = path.relative_path_from(@root).to_s
    stdout, status = Open3.capture2("git", "log", "-1", "--format=%ct", "--", relative, chdir: @root.to_s)
    return stdout.to_i if status.success? && !stdout.strip.empty?

    path.mtime.to_i
  end

  def match(content, regex)
    content.match(regex)&.captures&.first
  end
end

begin
  ReadmeUpdater.new(root: HomebrewTap.repo_root, dry_run: options.fetch(:dry_run)).run
rescue StandardError => e
  warn "Update failed: #{e.message}"
  exit 1
end
